app/services/journal_row_builder.rb

```diff
@@ -0,0 +1,165 @@
+require "set"
+
+# Ported from Journal model.
+# We should improve upon this logic.
+class JournalRowBuilder
+  include NUCore::Database::ArrayHelper
+
+  attr_reader :order_details, :journal, :journal_rows, :errors, :options,
+              :recharge_enabled, :product_recharges, :journaled_facility_ids
+
+  def initialize(journal, order_details)
+    @order_details = order_details
+    @journal = journal
+    reset
+  end
+
+  # Reset instance variables both on initialize and build.
+  def reset
+    @errors = []
+    @journal_rows = []
+    @journaled_facility_ids = Set.new
+    @product_recharges = {}
+    @recharge_enabled = SettingsHelper.feature_on?(:recharge_accounts)
+  end
+
+  # Idempotent method that builds an array of new journal_rows. Also adds to
+  # errors if any problems occur. Does not save anything to the database.
+  # Returns self to support chaining.
+  def build
+    reset
+    order_details.each do |order_detail|
+      validate_order_detail(order_detail)
+      new_journal_rows = order_detail_to_journal_rows(order_detail)
+      update_product_recharges(order_detail)
+      @journal_rows.concat(new_journal_rows)
```

Comment:
For the sake of keeping related logic together, is it possible to either combine `new_journal_rows = order_detail_to_journal_rows(order_detail)` and `@journal_rows.concat(new_journal_rows)` lines or put them next to each other instead of being separated by the `update_product_recharges(order_detail)` line.